### ISSUE TITLE
[Proposal] Git branching

### ISSUE THREAD

**Issue by tugcanolgun** [swe-ms-boun/2018fall-swe574-g1#8]
For git branching conventions, the contents of the following wiki link can be used.

[https://github.com/swe-ms-boun/2018fall-swe574-g1/wiki/Git-Branching](Git-Branching)

Please state if you have any objections, suggestions or if you simply approve this branching method.

**Comment by taneresme** [comment]
I saw your document in the wiki, your idea is great and I like it.
I have a few suggestions:
* We can put all release branches under a ``release`` folder like ``feature`` and ``fix``
* When we are merging the feature branch to ``develop`` and ``master``, we can create a pull request from ``feature`` branch to ``develop``. In this way, we will be able to correct some issues related to code quality and someone else in the team can have know-how about what you did in your ``feature``

**Comment by taneresme** [comment]
@tugcanolgun @vedatakses @macukadam @alidursen 
I have another recommendation for branching. We can fork separate repos from the repos in this group..
after completing each development, you can create a ``pull request`` to merge your feature branch to ``develop`` or ``master``. It will able to make us more flexible.

What do you think?

**Comment by tugcanolgun** [comment]
In the document, `release` branches are temporary. `fix`es and `feature`s are merged to `develop` branch. `develop` branch "manually" tested and then `release` branch is created. After this merge to the `master` is done via Pull Request and at least 1 (we may change this) other person has to accept this pull request. Then Travis will take over and do the unit tests. If the tests are successful, then the project will have a new version. With hooks, we can ensure the live deployment will get the new version as soon as the merge is complete. 

I believe the current system is sufficient for regular flow of this project. But we can of course change this behavior according to your and other member's suggestions.

**Comment by tugcanolgun** [comment]
> @tugcanolgun @vedatakses @macukadam @alidursen
> I have another recommendation for branching. We can fork separate repos from the repos in this group..
> after completing each development, you can create a `pull request` to merge your feature branch to `develop` or `master`. It will able to make us more flexible.
> 
> What do you think?

It may work. But it never should touch `master` branch. We can make pull requests to `develop` only. After we are all satisfied with the current state, we can merge it with release branch to `master`.

**Comment by taneresme** [comment]
> In the document, `release` branches are temporary. `fix`es and `feature`s are merged to `develop` branch. `develop` branch "manually" tested and then `release` branch is created. After this merge to the `master` is done via Pull Request and at least 1 (we may change this) other person has to accept this pull request. Then Travis will take over and do the unit tests. If the tests are successful, then the project will have a new version. With hooks, we can ensure the live deployment will get the new version as soon as the merge is complete.
> 
> I believe the current system is sufficient for regular flow of this project. But we can of course change this behavior according to your and other member's suggestions.

Fine @tugcanolgun, I just thought we were creating the release branch as making a tag for the releases... Maybe we need tags for releases to review what we released in each release when needed...

Anyway, It would be better if we talked in the next meeting together with others.

**Comment by taneresme** [comment]
> > @tugcanolgun @vedatakses @macukadam @alidursen
> > I have another recommendation for branching. We can fork separate repos from the repos in this group..
> > after completing each development, you can create a `pull request` to merge your feature branch to `develop` or `master`. It will able to make us more flexible.
> > What do you think?
> 
> It may work. But it never should touch `master` branch. We can make pull requests to `develop` only. After we are all satisfied with the current state, we can merge it with release branch to `master`.

I agree with you, We never touch ``master`` branch... We are on the right track :)

**Comment by vedatakses** [comment]
Do we really need a release branch? I would propose master and development branches. We can create a separate branch from development regarding the task we're working on and create a pull request to the development branch for 1-2 people. And maybe in every 2 weeks, we can merge the development branch to master, to keep master updated.

**Comment by tugcanolgun** [comment]
> Do we really need a release branch? I would propose master and development branches. We can create a separate branch from development regarding the task we're working on and create a pull request to the development branch for 1-2 people. And maybe in every 2 weeks, we can merge the development branch to master, to keep master updated.

We don't necessarily need it. If you guys think we should do it other way, I can update the branching page accordingly. I want to share with you all the article this schema is based on: https://nvie.com/posts/a-successful-git-branching-model/

**Comment by tugcanolgun** [comment]
I am marking this as accepted.
